File: src/com/fictional_particle_sim/geometrics/Vector.java
package com.fictional_particle_sim.geometrics;

import java.awt.*;


public class Vector {
    public Point start, end;
    public Color color;

    public Vector() {
        start = new Point();
        end= new Point();
    }
    public Vector(Point end) {
        this.start = new Point();
        this.end = end;
    }
    public Vector(Point end, Color color) {
        this.start = new Point();
        this.end = end;
        this.color = color;
    }
    public Vector(Point start, Point end) {
        this.start = start;
        this.end = end;
    }
    public Vector(Point start, Point end, Color color) {
        this.start = start;
        this.end = end;
        this.color = color;
    }
    public Vector(double r, double theta) {
        this.start = new Point();

        this.end = new Point(r * Math.cos(theta), r * Math.sin(theta));
    }

    public BoundingBox boundingBox() {

        return new BoundingBox(this);
    }

    public void setColor(Color color) {
        this.color = color;
    }

    public Vector center() {

        return new Vector(new Point(), this.end.sub(this.start), this.color);
    }
    public Vector center(Point c) {
        return this.center().shift(c);
    }

    public double angle() {
        Vector v1 = this.center();

        return Math.atan2(v1.end.y, v1.end.x);
    }
    public double magnitude() {

        return Math.hypot(this.center().end.x, this.center().end.y);
    }
    public Vector norm() {
        return this.scalar(1/this.magnitude());
    }

    public Vector scalar(double b) {

        return new Vector(this.center().end.mult(b), this.color).shift(this.start);
    }
    public Vector scaleFromOrigin(double b) {

        return new Vector(this.start.mult(b), this.end.mult(b), this.color);
    }

    public Vector shift(Point b) {
        return new Vector(this.start.add(b), this.end.add(b), this.color);
    }

    public Vector add(Vector b) {
        Vector v1 = this.center();
        Vector v2 = b.center();

        Vector temp = new Vector(v1.end.add(v2.end));
        return temp.shift(this.start);
    }
    public Vector sub(Vector b) {

        return this.add(b.scalar(-1));
    }

    public Vector[] comp() {
        Vector x = new Vector(new Point(this.start.x, this.start.y), new Point(this.end.x, this.start.y));
        Vector y = new Vector(new Point(this.start.x, this.start.y), new Point(this.start.x, this.end.y));
        x.setColor(Color.RED);
        y.setColor(Color.GREEN);

        return new Vector[]{x, y};
    }
    public Vector[] comp(boolean xToY) {
        Vector x = new Vector(new Point(this.start.x, this.start.y), new Point(this.end.x, this.start.y));
        Vector y = new Vector(new Point(this.start.x, this.start.y), new Point(this.start.x, this.end.y));
        x.setColor(Color.RED);
        y.setColor(Color.GREEN);
        if (xToY) {
            y = y.shift(new Point(this.center().end.x, 0));
        } else {
            x = x.shift(new Point(0, this.center().end.y));
        }

        return new Vector[]{x, y};
    }

    public boolean equals(Vector b) {
        Vector v1 = this.center();
        Vector v2 = b.center();

        return v1.end.equals(v2.end);
    }

    public double dot(Vector b) {
        Vector v1 = this.center();
        Vector v2 = b.center();

        return v1.end.x * v2.end.x + v1.end.y * v2.end.y;
    }
    public Vector proj(Vector b) {

        return b.norm().scalar(this.dot(b.norm())).center().shift(this.start);
    }
    public Vector perp(Vector b) {

        return this.sub(this.proj(b));
    }
    public Vector[] projComp(Vector b) {
        Vector x = this.proj(b);
        Vector y = this.perp(b);

        x.setColor(Color.RED);
        y.setColor(Color.GREEN);

        return  new Vector[]{x, y};
    }
    public Vector[] projComp(Vector b, boolean xToY) {
        Vector x = this.proj(b);
        Vector y = this.perp(b);
        x.setColor(Color.RED);
        y.setColor(Color.GREEN);

        if (xToY) {
            y = y.shift(x.center().end);
        } else {
            x = x.shift(y.center().end);
        }

        return  new Vector[]{x, y};
    }

    public Vector rotate(double angle) {

        return new Vector(this.magnitude(), this.angle() + angle).shift(this.start);
    }

    public boolean between(Point a, Point b) {
        Vector ab = new Vector(a, b);
        if (ab.boundingBox().intersect(boundingBox())) {
            return linesIntersect(this, ab);
        } else return false;
    }

    private static boolean linesIntersect(Vector v1, Vector v2) {
        double x1 = v1.start.x;
        double y1 = v1.start.y;
        double x2 = v1.end.x;
        double y2 = v1.end.y;

        double x3 = v2.start.x;
        double y3 = v2.start.y;
        double x4 = v2.end.x;
        double y4 = v2.end.y;

        if (x1 == x2 && y1 == y2 || x3 == x4 && y3 == y4) {
            return false;
        }

        double ax = x2 - x1;
        double ay = y2 - y1;
        double bx = x3 - x4;
        double by = y3 - y4;
        double cx = x1 - x3;
        double cy = y1 - y3;

        double alphaNumerator = by * cx - bx * cy;
        double commonDenominator = ay * bx - ax * by;
        if (commonDenominator > 0) {
            if (alphaNumerator < 0 || alphaNumerator > commonDenominator) {
                return false;
            }
        } else if (commonDenominator < 0) {
            if (alphaNumerator > 0 || alphaNumerator < commonDenominator) {
                return false;
            }
        }
        double betaNumerator = ax * cy - ay * cx;
        if (commonDenominator > 0) {
            if (betaNumerator < 0 || betaNumerator > commonDenominator) {
                return false;
            }
        } else if (commonDenominator < 0) {
            if (betaNumerator > 0 || betaNumerator < commonDenominator) {
                return false;
            }
        }
        if (commonDenominator == 0) {

            double y3LessY1 = y3 - y1;
            double collinearityTestForP3 = x1 * (y2 - y3) + x2 * (y3LessY1) + x3 * (y1 - y2);

            if (collinearityTestForP3 == 0) {
                if (x1 >= x3 && x1 <= x4 || x1 <= x3 && x1 >= x4 || x2 >= x3 && x2 <= x4 || x2 <= x3 && x2 >= x4 || x3 >= x1 && x3 <= x2 || x3 <= x1 && x3 >= x2) {
                    return y1 >= y3 && y1 <= y4 || y1 <= y3 && y1 >= y4 || y2 >= y3 && y2 <= y4 || y2 <= y3 && y2 >= y4 || y3 >= y1 && y3 <= y2 || y3 <= y1 && y3 >= y2;
                }
            }
            return false;
        }
        return true;
    }

    public boolean opposite(Vector b) {

        return this.norm().equals(b.norm().scalar(-1));
    }

    @Override
    public String toString() {
        return center().end.toString();
    }
}

File: src/com/fictional_particle_sim/geometrics/Vector.kt
package com.fictional_particle_sim.geometrics

import java.awt.Color
import kotlin.math.*

class Vector(var start: Point = Point(), var end: Point = Point(), var color: Color = Color.WHITE) {

    companion object{
        @JvmStatic
        fun vector(r: Double, theta: Double, start: Point = Point(), color: Color = Color.WHITE) = Vector(start, Point(r * cos(theta), r * sin(theta)) + start, color)
    }

    fun boundingBox(): BoundingBox {
        val a = Point(start.x.coerceAtMost(end.x), start.y.coerceAtMost(end.y))
        val b = Point(start.x.coerceAtLeast(end.x), start.y.coerceAtLeast(end.y))
        return BoundingBox(a, b)
    }

    operator fun plus(that: Vector) = Vector(end = center().end + that.center().end, color = color).center(start)
    operator fun plus(that: Point) = Vector(start + that, end + that, color)

    operator fun minus(that: Vector) = this + -that

    operator fun times(that: Double) = Vector(end = center().end * that, color = color).center(start)
    operator fun times(that: Int) = this * that.toDouble()
    operator fun Double.times(that: Vector) = that * this
    operator fun Int.times(that: Vector) = that * this

    fun scaleFromOrigin(that: Double) = Vector(start * that, end * that, color)

    operator fun div(that: Double) = this * (1.0/that)

    operator fun unaryPlus() = this
    operator fun unaryMinus() = this * -1

    fun x() = center().end.x
    fun y() = center().end.y

    fun center(center: Point = Point()) = Vector(center, end - start + center, color)

    fun angle() = atan2(y(), x())
    fun magnitude() = hypot(x(), y())
    fun norm() = this / magnitude()

    fun comp(): Array<Vector> {
        val x = Vector(Point(start.x, start.y), Point(end.x, start.y), Color.RED)
        val y = Vector(Point(start.x, start.y), Point(start.x, end.y), Color.GREEN)

        return arrayOf(x, y)
    }

    fun dot(that: Vector): Double = x() * that.x() + y() * that.y()
    fun proj(that: Vector) = (that.norm() * (dot(that.norm()))).center(start)
    fun perp(that: Vector) = this - proj(that)
    fun projComp(that: Vector): Array<Vector> {
        val x = proj(that)
        val y = perp(that)
        x.color = Color.RED
        y.color = Color.GREEN

        return arrayOf(x, y)
    }

    fun rotate(theta: Double) = vector(magnitude(), angle() + theta, start, color)

    fun between(a: Point, b: Point): Boolean {
        val ab: Vector = Vector(a, b)
        return if (ab.boundingBox().intersect(boundingBox())) {
            linesIntersect(this, ab)
        } else false
    }

    private fun linesIntersect(v1: Vector, v2: Vector): Boolean {
        val x1 = v1.start.x
        val y1 = v1.start.y
        val x2 = v1.end.x
        val y2 = v1.end.y

        val x3 = v2.start.x
        val y3 = v2.start.y
        val x4 = v2.end.x
        val y4 = v2.end.y

        if (x1 == x2 && y1 == y2 || x3 == x4 && y3 == y4) {
            return false
        }

        val ax = x2 - x1
        val ay = y2 - y1
        val bx = x3 - x4
        val by = y3 - y4
        val cx = x1 - x3
        val cy = y1 - y3

        val alphaNumerator = by * cx - bx * cy
        val commonDenominator = ay * bx - ax * by
        if (commonDenominator > 0) {
            if (alphaNumerator < 0 || alphaNumerator > commonDenominator) {
                return false
            }
        } else if (commonDenominator < 0) {
            if (alphaNumerator > 0 || alphaNumerator < commonDenominator) {
                return false
            }
        }
        val betaNumerator = ax * cy - ay * cx
        if (commonDenominator > 0) {
            if (betaNumerator < 0 || betaNumerator > commonDenominator) {
                return false
            }
        } else if (commonDenominator < 0) {
            if (betaNumerator > 0 || betaNumerator < commonDenominator) {
                return false
            }
        }
        if (commonDenominator == 0.0) {
            val y3LessY1 = y3 - y1
            val collinearityTestForP3 = x1 * (y2 - y3) + x2 * y3LessY1 + x3 * (y1 - y2)
            if (collinearityTestForP3 == 0.0) {
                if (x1 >= x3 && x1 <= x4 || x1 <= x3 && x1 >= x4 || x2 >= x3 && x2 <= x4 || x2 <= x3 && x2 >= x4 || x3 >= x1 && x3 <= x2 || x3 <= x1 && x3 >= x2) {
                    return y1 >= y3 && y1 <= y4 || y1 <= y3 && y1 >= y4 || y2 >= y3 && y2 <= y4 || y2 <= y3 && y2 >= y4 || y3 >= y1 && y3 <= y2 || y3 <= y1 && y3 >= y2
                }
            }
            return false
        }
        return true
    }

    override fun toString(): String = center().end.toString()
    fun toString(inPolar: Boolean): String = if (inPolar) {
            "( r = ${magnitude()} , theta = ${angle()} )"
        } else toString()

}